Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp

```diff
@@ -77,7 +78,135 @@ Job* WorkQueue::TryStealFront()
+    }
+
+    [[nodiscard]] bool empty() const
+    {
+        return head == tail;
+    }
+
+    // Typical circular buffer pattern where the head chases the tail. For two phase commits, tailReserve can run ahead of tail
+    // but can't bump into head. Note that for all operations here, we rely on well-defined wrapping behavior of unsigned ints.
+    [[nodiscard]] bool canEnqueue() const
+    {
+        return tailReserve != head - 1;
+    }
+
+    void reserve()
+    {
+        ++tailReserve;
+    }
+
+    void commit()
+    {
+        ++tail;
+    }
+
+    void acquire()
+    {
+        ++head;
+    }
+
+    uint16_t head;
+    uint16_t tail;
+    uint16_t tailReserve;
+    uint16_t unused;
+};
+
+void JobQueue::Enqueue(Job* job)
+{
+    uint8_t priority = job->GetPriority();
+    uint64_t previousStatus = m_status[priority].load(AZStd::memory_order_acquire);
+
+    while (true)
+    {
+        StatusPayload status{ previousStatus };
+
+        // Enqueuing is done in two phases because we cannot atomically write the job to the slot we reserve
+        // and simulataneously publish the fact that the slot is now available.
+        if (status.canEnqueue())
+        {
+            // Try to reserve a slot
+            status.reserve();
+            uint64_t reserveStatus = status.pack();
+            if (m_status[priority].compare_exchange_weak(previousStatus, reserveStatus))
+            {
+                m_queues[priority][status.tailReserve - 1] = job;
+
+                status.commit();
+
+                while (true)
+                {
+                    // Increment the tail to advertise the new job
+                    if (m_status[priority].compare_exchange_weak(reserveStatus, status.pack()))
+                    {
+                        return;
+                    }
+
+                    status = reserveStatus;
+                    status.commit();
+                }
+            }
+
+            // We failed to reserve a slot, try again
+        }
+        else
+        {
+            // All queues at or below the requested priority are full. Queue in the global unbounded queue instead.
+            job->GetContext()->GetJobManager().QueueUnbounded(job);
+            job->GetContext()->GetJobManager().ActivateWorker();
+            return;
+        }
+    }
+}
+
+Job* JobQueue::TryDequeue()
+{
+    for (size_t priority = 0; priority != PRIORITY_LEVEL_COUNT; ++priority)
+    {
+        uint64_t previousStatus = m_status[priority].load(AZStd::memory_order_acquire);
+        while (true)
+        {
+            StatusPayload status{ previousStatus };
+
+            if (status.empty())
+            {
+                // Queue empty
+                break;
+            }
+            else
+            {
+                status.acquire();
+                if (m_status[priority].compare_exchange_weak(previousStatus, status.pack()))
+                {
+                    return m_queues[priority][status.head - 1];
+                }
```

Comment:
This is technically a race condition. It's possible, though unlikely that the `compare_exchange_weak` releases the slot in the queue, it's picked up by another thread and that thread writes a new job in it, which will cause this function to return an invalid job pointer. Instead the job needs to be stored before the `compare_exchange_weak` and the cached version should be returned.

Reply:
No this should be protected by the two phase enqueue above. Can you describe the exact sequence of instructions  you believe demonstrates the issue?

Reply:
When this line `if (m_status[priority].compare_exchange_weak(previousStatus, status.pack()))` is successful the specific slot `[status.head-1]` is available again as far as the rest of the threads are concerned, which means that any thread can now use that slot to write a job to. However this thread isn't really done with that slot yet as it still needs to read the job address from it, so the follow (very unlikely) situation could happen:
```
status.acquire();
if (m_status[priority].compare_exchange_weak(previousStatus, status.pack()))
{
    // This thread is suspended for a long time and all other threads in the meantime collectively push another 65536
    // jobs to this queue. Since the cas already released the slot that number 65536 will be the last that can be written
    // but will overwrite this slot so a different job than intended is returned.
    return m_queues[priority][status.head - 1];
}
```
The two step queuing doesn't protect against that because that only guards against another thread picking the same slot while the first thread is still busy writing the job pointer. This is due to the thread owning the queue releasing a slot and then reading it instead of the other way around.
I might be wrong here again, since I also embarrassingly forget `exchange_compare_weak` returns the actual value, but if not then the fix is trivial:
```
status.acquire();
Job* result = m_queues[priority][status.head - 1];
if (m_status[priority].compare_exchange_weak(previousStatus, status.pack()))
{
    return result;
}
```

Reply:
Hmm not quite. If the compare exchange is successful, we've already incremented `head` to the next slot. Another thread is free to take that next slot at the same time if it wants. Also, the queue can never be full because once the queue is empty, we immediately enqueue to the overflow queue (which is unbounded and guarded by a lock).

Reply:
Fixed thanks!